etcd/src/test/scala/io/buoyant/etcd/KeyTest.scala

```diff
@@ -0,0 +1,441 @@
+      val (path, params) = req.method match {
+        case Method.Post | Method.Put =>
+          val path = Path.read(req.uri)
+          val params = FormParams.get(req)
+          (path, params)
+
+        case _ =>
+          val path = Path.read(req.path)
+          val params = req.params.keys.foldLeft(Map.empty[String, Seq[String]]) {
+            case (params, k) => params + (k -> req.params.getAll(k).toSeq)
+          }
+          (path, params)
+      }
+
+      val rsp = Response()
+      rsp.version = req.version
+
+      val k = (req.method, path, params)
+      if (handle.isDefinedAt(k)) {
+        Future(handle(k)).flatMap(serve(req.method, rsp, _)) handle {
+          case e@ApiError(code, _, _, index) =>
+            rsp.status = code match {
+              case ApiError.KeyNotFound => Status.NotFound
+              case ApiError.NodeExist => Status.Forbidden
+              case _ => Status.BadRequest
+            }
+            addState(Etcd.State(index), rsp)
+            rsp.content = writeJson(e)
+            rsp
+        }
+      } else {
+        rsp.status = Status.InternalServerError
+        info(s"method=${req.method} path=$path params=$params")
+        Future.value(rsp)
+      }
+    }
+
+  private[this] def mkEtcd(
+    filter: Filter[Request, Response, Request, Response] = Filter.identity
+  )(handle: PartialFunction[(Method, Path, Params), Any]) =
+    new Etcd(filter andThen mkClient(handle))
+
+  private[this] def mkKey(
+    key: Path,
+    filter: Filter[Request, Response, Request, Response] = Filter.identity
+  )(handle: PartialFunction[(Method, Params), Any]): Key = {
+    val uri = Path.Utf8("v2", "keys") ++ key
+    val etcd = mkEtcd(filter) {
+      case (method, `uri`, params) if handle.isDefinedAt(method, params) =>
+        handle((method, params))
+    }
+    etcd.key(key)
+  }
+
+  private[this] def serve(method: Method, rsp: Response, v: Any): Future[Response] =
+    v match {
+      case f: Future[_] =>
+        f.flatMap(serve(method, rsp, _))
+
+      case Some(v) =>
+        serve(method, rsp, v)
+
+      case op: NodeOp =>
+        rsp.content = writeJson(NodeOp.Json(op))
+        addState(op.etcd, rsp)
+        rsp.status = (rsp.status, method) match {
+          case (Status.Ok, Method.Put | Method.Post) => Status.Created
+          case (status, _) => status
+        }
+        Future.value(rsp)
+
+      case v: Version =>
+        rsp.content = writeJson(v)
+        Future.value(rsp)
+
+      case idk =>
+        rsp.status = Status.InternalServerError
+        rsp.contentString = idk.toString
+        Future.value(rsp)
+    }
+
+  private[this] def addState(etcd: Etcd.State, msg: Message): Unit = {
+    msg.headerMap("X-Etcd-Index") = etcd.index.toString
+    msg.headerMap("X-Etcd-Cluster-Id") = etcd.clusterId
+  }
+
+  private[this] object Recursive {
+    def unapply(params: Params): Option[Boolean] =
+      Some(getParam(params, "recursive").exists(_ == "true"))
```

Comment:
tioli, code-golf: `contains("true")` is slightly fewer characters (and also slightly more efficient, not that matters here in a test)